lib/IRGen/GenFunc.cpp

```diff
@@ -2433,3 +2434,83 @@ IRGenFunction::createAsyncDispatchFn(const FunctionPointer &fnPtr,
   Builder.CreateRetVoid();
   return dispatch;
 }
+
+void IRGenFunction::emitSuspensionPoint(llvm::Value *toExecutor,
+                                        llvm::Value *asyncResume) {
+  // TODO: pointerauth
+
+  // Setup the suspend point.
+  SmallVector<llvm::Value *, 8> arguments;
+  arguments.push_back(asyncResume);
+  auto resumeProjFn = getOrCreateResumeFromSuspensionFn();
+  arguments.push_back(
+      Builder.CreateBitOrPointerCast(resumeProjFn, IGM.Int8PtrTy));
+  llvm::Function *suspendFn = createAsyncSuspendFn();
+  arguments.push_back(
+      Builder.CreateBitOrPointerCast(suspendFn, IGM.Int8PtrTy));
+
+  arguments.push_back(asyncResume);
+  arguments.push_back(
+      Builder.CreateBitOrPointerCast(toExecutor, getAsyncExecutor()->getType()));
+  arguments.push_back(getAsyncTask());
+  arguments.push_back(getAsyncExecutor());
+  arguments.push_back(getAsyncContext());
+
+  emitSuspendAsyncCall(arguments);
+}
+
+llvm::Function *IRGenFunction::getOrCreateResumeFromSuspensionFn() {
+  auto name = "__swift_async_resume_get_context";
+  return cast<llvm::Function>(IGM.getOrCreateHelperFunction(
+      name, IGM.Int8PtrTy, {IGM.Int8PtrTy},
+      [&](IRGenFunction &IGF) {
+        auto &Builder = IGF.Builder;
+        Builder.CreateRet(&*IGF.CurFn->arg_begin());
+      },
+      false /*isNoInline*/));
+}
```

Comment:
I'm a little confused about what this function does.  It looks to be avoiding assuming stuff about the resumption function signature, but don't we assume a bunch of stuff about function signatures in the async lowering anyway?  Or have we managed to avoid that so far?

Reply:
An async suspend point needs to know how-to restore the current context when resuming from the suspend point (how-to get from the resume function's context argument to the current context when the continuation function is called). This is described as a function pointer argument to the async suspend intrinsic; LLVM's async lowering guarantees inlining of this function.

Another example of this: A suspension point that models an async call restores the current context by loading the parent context from the callee context (the current implementation does caller pop).

Reply:
> Or have we managed to avoid that so far?
Yes. LLVM's async lowering tries to make as little assumption about the signature as possible. The `llvm.coro.id.async` intrinsic describes at which argument index the context argument resides. The `llvm.coro.suspend.async` intrinsic describes how-to restore the current context from the resume function's context argument.